### PR TITLE
Transfer - Double value progress bar

### PR DESCRIPTION
**Pull request by sarao1310**
Signed-off-by: sarao <sarao@jfrog.com>

- [ ] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [ ] This pull request is on the dev branch.
- [ ] I used gofmt for formatting the code before submitting the pull request.
-----
  providing double value progress bar that give the user more information about transferred storage and files

### CODE REVIEW

**Comment by sarao1310** [comment]
I have read the CLA Document and I hereby sign the CLA

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 1]
```diff
@@ -112,6 +127,10 @@ func (t *TransferProgressMng) Quit() error {
 		if t.totalRepositories != nil {
 			t.barsMng.QuitTasksWithHeadlineProg(t.totalRepositories)
 		}
+		if t.totalSize != nil {
+			t.totalSize.GetBar().Abort(true)
+			t.totalSize = nil
+		}
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/transferfileprogress.go:133]: Any reason why the above code lines aren't placed inside the `abortMetricsBars` function?
The rest of progress bars are aborted there.

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 2]
```diff
@@ -222,6 +241,17 @@ func (t *TransferProgressMng) RemoveRepository() {
...
 
+func (t *TransferProgressMng) RemoveFile(chunkSize int) {
+	if t.currentChunkHeadLine == nil {
+		return
+	}
+	t.IncreaseTotalSize(chunkSize)
+	t.currentChunkHeadLine.Abort(true)
+	t.emptyLine.Abort(true)
+	t.emptyLine = nil
+	//abort all phases bars
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/transferfileprogress.go:252]: This comment should be probably removed.

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 3]
```diff
@@ -222,6 +241,17 @@ func (t *TransferProgressMng) RemoveRepository() {
 	time.Sleep(progressbar.ProgressRefreshRate)
 }
 
+func (t *TransferProgressMng) RemoveFile(chunkSize int) {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/transferfileprogress.go:244]: The name of this function seem wrong and unclear. By reading the name of this function, the reader can this that it actually removes a file. 
Actually, this function seem to be unused. Should it be removed?

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 4]
```diff
@@ -235,6 +265,10 @@ func (t *TransferProgressMng) SetRunningThreads(n int) {
 	}
 }
 
+func (t *TransferProgressMng) IncreaseTotalSize(n int) {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/transferfileprogress.go:268]: Looks like this function can be made provate.

File: artifactory/commands/transferfiles/utils.go [hunk 5]
```diff
@@ -295,6 +295,7 @@ func handleChunksStatuses(phase *phaseBase, chunksStatus *UploadChunksStatusResp
 }
 
 func updateProgress(phase *phaseBase, progressbar *TransferProgressMng, timeEstMng *timeEstimationManager, chunk ChunkStatus, workingThreads int) error {
+	log.Info("I'm in updateProgress")
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:298]: Should be probably removed.

File: artifactory/commands/transferfiles/utils.go [hunk 6]
```diff
@@ -295,6 +295,7 @@ func handleChunksStatuses(phase *phaseBase, chunksStatus *UploadChunksStatusResp
 }
 
 func updateProgress(phase *phaseBase, progressbar *TransferProgressMng, timeEstMng *timeEstimationManager, chunk ChunkStatus, workingThreads int) error {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:297]: Let'a add a description for this fuunction.

File: utils/progressbar/progressbarmng.go [hunk 7]
```diff
@@ -194,11 +194,25 @@ func (bm *ProgressBarMng) NewStringProgressBar(headline string, updateFn func()
...
 		mpb.PrependDecorators(
 			decor.Name(headline),
-			decor.Any(func(statistics decor.Statistics) string {
-				return updateFn()
+		),
+	)
+	return pb
+}
+
+func (bm *ProgressBarMng) NewDoubleValuesProgressBar(headLine string, color Color, totalTasks int64, updateFunc func() string) *TasksProgressBar {
```
Comment (eyalbe4) [inline utils/progressbar/progressbarmng.go:202]: NewDoubleValuesProgressBar --> NewDoubleValueProgressBar

Comment (eyalbe4) [inline utils/progressbar/progressbarmng.go:202]: Let's add a clear description for this function, expalining what `DoubleValueProgressBar` actually is.
Let's do the same for the `NewStringProgressBar`, `NewCounterProgressBar`, `NewDoneTasksProgressBar` and `NewTasksProgressBar` functions. This should really help the reader understand the differences between them.

Comment (eyalbe4) [inline utils/progressbar/progressbarmng.go:202]: This function receives only one headline as an argument. The other one ("Files") is hard coded inside the function, and isn't received as an argument, which make the usage a bit strange.
Let's have have the first two arguments be `headLine1,  headLine2 string`

This function isn't really generic. Meaning, it doesn't really allow you to create any type of "double value progress" you wish. It is very specific to our use case, which is showing the total of files and total storage. For example, the function specifically uses `"%.1f/%.1f"`, which is good for our use case only.
I see two options to design this properly - 
1. Either make the function truely generic, allowing the user to configure it to show any two values.
2. Keep it specific for our use case, but name it "NewTotalFilesAndStorageProgressBar". Also In this case, it doesn't make sense that the update logic will be separated into an `updateFunc()` that is sent as an argument, becuase this progress cannot be customized anyway.

File: utils/progressbar/taskswithheadlineprog.go [hunk 8]
```diff
@@ -10,6 +10,12 @@ type TasksWithHeadlineProg struct {
 	tasksProgressBar *TasksProgressBar
 	emptyLine        *mpb.Bar
 }
+type TasksWithTwoHeadLinesProg struct {
```
Comment (eyalbe4) [inline utils/progressbar/taskswithheadlineprog.go:13]: 1. Missing new line before the declaration of this struct.
2. This struct doesn't include a description like the the struct above it. Maintaining consistency in the code is very important.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I'd be happy to review this PR again after implementing or discussing my comments.

File: artifactory/commands/transferfiles/transfer.go [hunk 9]
```diff
@@ -131,7 +131,7 @@ func (tdc *TransferFilesCommand) Run() (err error) {
 	}
 
 	// Set progress bar with the length of the source local and build info repositories
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply artifactory/commands/transferfiles/transfer.go:133]: Done!

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 10]
```diff
@@ -30,6 +31,7 @@ type TransferProgressMng struct {
 	shouldDisplay bool
 	// Task bar with the total repositories transfer progress
 	totalRepositories *progressbar.TasksWithHeadlineProg
+	totalSize         *progressbar.TasksProgressBar
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply artifactory/commands/transferfiles/transferfileprogress.go:34]: Done!

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 11]
```diff
@@ -45,39 +47,51 @@ type TransferProgressMng struct {
...
 
 // NewTransferProgressMng creates TransferProgressMng object.
 // If the progress bar shouldn't be displayed returns nil.
-func NewTransferProgressMng(allSourceLocalRepos []string, timeEstMng *timeEstimationManager, ignoreOldState bool) (*TransferProgressMng, error) {
+func NewTransferProgressMng(allSourceLocalRepos []string, tdc *TransferFilesCommand, fileStatus int) error {
 	totalRepositories := int64(len(allSourceLocalRepos))
+	totalSize, err := tdc.sourceStorageInfoManager.GetReposTotalSize(allSourceLocalRepos...)
+	storageInfo, err := tdc.sourceStorageInfoManager.GetStorageInfo()
+	totalFiles := storageInfo.BinariesCount
+	//Sara
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply artifactory/commands/transferfiles/transferfileprogress.go:64]: Done!

File: artifactory/commands/transferfiles/utils.go (hunk 6 above)
Comment (sarao1310) [reply artifactory/commands/transferfiles/utils.go:297]: Done!

File: utils/progressbar/progressbarmng.go (hunk 7 above)
Comment (sarao1310) [reply utils/progressbar/progressbarmng.go:202]: Done!

Comment (sarao1310) [reply utils/progressbar/progressbarmng.go:202]: Done!

File: utils/progressbar/taskswithheadlineprog.go (hunk 8 above)
Comment (sarao1310) [reply utils/progressbar/taskswithheadlineprog.go:13]: This struct is not needed so I deleted it.

File: utils/progressbar/progressbarmng.go (hunk 7 above)
Comment (sarao1310) [reply utils/progressbar/progressbarmng.go:202]: 1. I updated the input to two headlines
2. all the progress bars c'tors used a specific updating format in the body of the c'tor, I also think that is more generic and elegant to make the showing update format more generic but then I think that we have to change all the progress bars c'tors.. what you think Eyal? 
3. the updateFunc() is not an option, it's one of the conditions to use (Any) decorate in the bar library so we just added it

**Review by yahavi — CHANGES_REQUESTED** [review]
Following https://github.com/jfrog/jfrog-cli-core/pull/581, let's do the following:
1. Add TransferredFiles and TotalFiles to TransferRunStatus. Populate them in `updateProgress`.
2. Add transferred files information in `status.go`.
3. Optional - Export sizeToString from `status.go` and use it.

File: utils/progressbar/progressbarmng.go (hunk 7 above)
Comment (sarao1310) [reply utils/progressbar/progressbarmng.go]: It is very generic now

File: artifactory/commands/transferfiles/transferfileprogress.go [hunk 12]
```diff
@@ -45,39 +47,51 @@ type TransferProgressMng struct {
...
 	// In case of an emergency stop the transfer's progress bar will be aborted and the 'stopLine' bar will be display.
-	stopLine *mpb.Bar
+	stopLine    *mpb.Bar
+	filesStatus int
 }
 
 // NewTransferProgressMng creates TransferProgressMng object.
 // If the progress bar shouldn't be displayed returns nil.
-func NewTransferProgressMng(allSourceLocalRepos []string, timeEstMng *timeEstimationManager, ignoreOldState bool) (*TransferProgressMng, error) {
+func NewTransferProgressMng(allSourceLocalRepos []string, tdc *TransferFilesCommand, fileStatus int) error {
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply artifactory/commands/transferfiles/transferfileprogress.go]: Done!

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I haven'r reviewed the code yet. Let me do it soon.
In the meantime, I'm sharing the following feedback.
1. Let’s put the overall Repositories progress at the top (above the overall Files & Storage progress).
2. Let’s remove the “Diff” prefix from the Phase 2 progress bar. The label can be the same as Phase 1.
3. The two overall progress bars should be white, and the Current Repository progress bars should be grenn. The color difference is meant to differentiate between the “overall” and “repository specific” progress bars.

**Comment by sarao1310** [comment]
> I haven'r reviewed the code yet. Let me do it soon. In the meantime, I'm sharing the following feedback.
> 
> 1. Let’s put the overall Repositories progress at the top (above the overall Files & Storage progress).
> 2. Let’s remove the “Diff” prefix from the Phase 2 progress bar. The label can be the same as Phase 1.
> 3. The two overall progress bars should be white, and the Current Repository progress bars should be grenn. The color difference is meant to differentiate between the “overall” and “repository specific” progress bars.

1 and 3 are Okay, But I don't agree with you to remove "Diff" prefix from the phase 2 pregress bar, because I think the "Diff" prefix gives the user a notion about the job of phase 2

File: utils/progressbar/progressbarmng.go [hunk 13]
```diff
@@ -194,14 +205,53 @@ func (bm *ProgressBarMng) NewStringProgressBar(headline string, updateFn func()
...
-				return updateFn()
-			}),
 		),
 	)
 	return pb
 }
 
+// A progress bar with two counters values shown on the right side of the progress bar, the first value controls what the bar shows
+// The total tasks1 can be passes by an int or *int, if you want to use it with int send nil to the optional total and done tasks1 and the wanted totalTasks to total tasks1
+// else if you want to use
```
Comment (eyalbe4) [inline utils/progressbar/progressbarmng.go:215]: Let's rephrase and use proper punctionation for this description, to make it easier to read and understand.

File: utils/progressbar/progressbarmng.go [hunk 14]
```diff
@@ -83,6 +85,15 @@ func (bm *ProgressBarMng) NewTasksWithHeadlineProg(totalTasks int64, headline st
 	return &prog
 }
 
+func (bm *ProgressBarMng) NewHeadLineDoubleProgBar(headLine string, counterLine string, counterLine2 string, color Color, totalTasks1 int64, optionalTotal1, optionalDone1, totalTasks2, DoneTasks2 *int64) *TasksWithHeadlineProg {
```
Comment (eyalbe4) [inline utils/progressbar/progressbarmng.go:88]: Let's add a cealr descriotuo for this function, and also for all the other functions generating the various progress bars.

NewHeadLineDoubleProgBar --> NewHeadLineDoubleValProgressBar

File: utils/progressbar/progressbarmng.go [hunk 15]
```diff
@@ -194,14 +205,53 @@ func (bm *ProgressBarMng) NewStringProgressBar(headline string, updateFn func()
...
-			}),
 		),
 	)
 	return pb
 }
 
+// A progress bar with two counters values shown on the right side of the progress bar, the first value controls what the bar shows
+// The total tasks1 can be passes by an int or *int, if you want to use it with int send nil to the optional total and done tasks1 and the wanted totalTasks to total tasks1
+// else if you want to use
+func (bm *ProgressBarMng) NewDoubleValueProgressBar(headLine1 string, headLine2 string, color Color, totalTasks1 int64, OptionalTotalTasks1, optionalDoneTasks1, totalTasks2, doneTaks2 *int64) *TasksProgressBar {
```
Comment (eyalbe4) [inline utils/progressbar/progressbarmng.go:216]: Let's rename and reorder thre arguments list (for example, it makes more sense to have the color as the last argument).

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: artifactory/commands/transferfiles/transferfileprogress.go (hunk 4 above)
Comment (sarao1310) [reply artifactory/commands/transferfiles/transferfileprogress.go:264]: Done!

File: artifactory/commands/transferfiles/state/runstatus.go [hunk 16]
```diff
@@ -24,7 +24,8 @@ type ActionOnStatusFunc func(transferRunStatus *TransferRunStatus) error
 type TransferRunStatus struct {
 	lastSaveTimestamp time.Time
 	// This variable holds the total/transferred number of repositories (not their files).
-	TotalRepositories ProgressState      `json:"total_repositories,omitempty"`
+	TransferOverall   ProgressState      `json:"transfer_overall,omitempty"`
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply artifactory/commands/transferfiles/state/runstatus.go:27]: Done

File: artifactory/utils/storageinfo_test.go [hunk 17]
```diff
@@ -159,3 +159,22 @@ func getStorageInfoResponse(t *testing.T, w http.ResponseWriter, r *http.Request
...
+		{12546, "12.3KB "},
+		{148576, "145.1KB "},
+		{2587985, "2.5MB "},
+		{12896547, "12.3MB "},
+		{12896547785, "12.0GB "},
+		{5248965785422365, "4773.9TB "},
+	}
+
+	for _, test := range tests {
+		//actualCompName, actualCompVersion, actualCompType := splitComponentId(test.componentId)
```
Comment (eyalbe4) [inline artifactory/utils/storageinfo_test.go:177]: Let'e remove.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Pull request merged**